mapfishapp/src/main/java/org/georchestra/mapfishapp/ws/AddonController.java

```diff
@@ -0,0 +1,163 @@
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.context.ServletContextAware;
+
+@Controller
+public class AddonController implements ServletContextAware {
+
+    @Autowired
+    private GeorchestraConfiguration georchestraConfiguration;
+
+    private ServletContext context;
+
+    private static final Log LOG = LogFactory.getLog(AddonController.class.getPackage().getName());
+
+    /**
+     * Returns a dynamically-generated array of addons, previously
+     * statically stored into the GEOR-custom.js file.
+     *
+     * The returned format looks like:
+     * [{
+     *   "id": "magnifier_0", // unique & stable string identifier for this addon instance
+     *   "name": "Magnifier",
+     *   "title": {
+     *       "en": "Aerial imagery magnifier",
+     *       "es": "Lupa ortofoto",
+     *       "fr": "Loupe orthophoto",
+     *       "de": "Ortofoto Lupe"
+     *   },
+     *   "description": {
+     *       "en": "A tool which allows to zoom in an aerial image on a map portion",
+     *       "es": "Una herramienta que permite hacer un zoom sobre una parte del mapa ortofoto",
+     *       "fr": "Un outil qui permet de zoomer dans une orthophoto sur une portion de la carte",
+     *       "de": "Utensil erlaubt Zoom mittels orthofoto auf Kartenbereich"
+     *   }
+     *  }, ...]
+     *
+     * The returned JSON is constructed reading the config.json of each addons.
+     *
+     * @param request
+     * @param response
+     * @throws Exception
+     *
+     */
+    @RequestMapping(value= "/addons")
+    public void getAddons(HttpServletRequest request, HttpServletResponse response) throws Exception {
+        response.setContentType("application/javascript; charset=UTF-8");
+        JSONArray jsAddons = constructAddonsSpec();
+
+        response.getOutputStream().write(jsAddons.toString(4).getBytes());
+    }
+
+    @RequestMapping(value="/addons/**")
+    public void getAddonFile(HttpServletRequest request, HttpServletResponse response) throws Exception {
+        String pathInfo = request.getPathInfo().replaceFirst("/addons/", "");
+        String path = getMfappAddonPath();
+
+        Path reqPath = new File(path, pathInfo).toPath();
+        Path datadir = new File(path).toPath();
+
+        if (! reqPath.toAbsolutePath().toString().startsWith(datadir.toAbsolutePath().toString())) {
+            LOG.error("Unexpected file requested (not in datadir): " + reqPath.toString());
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            return;
+        }
+        File actualFile = new File(String.format("%s%s%s", datadir, File.separator, pathInfo));
+        if (! actualFile.exists()) {
+            LOG.error("requested file does not exist: " + actualFile);
+            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+            return;
+        }
+        String ext = FilenameUtils.getExtension(actualFile.getAbsolutePath());
```

Comment:
I've a basic check that ensures we actually are in the datadir (line 80), and I hardly remember having coded a unit test that "penetration tests" this piece of code. Of course, having some code that dumps file to the client without basic checks on what is requested is actually dangerous, hence my check on line 80. I think this should address the LFI (local file inclusion) risk, by jailing the possible files served into georchestra.datadir/mapfishapp/addons/